2/answers.txt

```diff
@@ -0,0 +1,231 @@
+
+Thread safety is about ensuring that data structure manipulated by multiple threads affect the data in the intended way, without causing unexpected changes or damage as a result of multiple entry points to the data structure.
+Typically these are done by implementing race conditions, however you can achieve thread safety without race conditions by implementing, Re-Entrancy, thread-local storage, immutable objects, mutual exclusion and atomic operations.
+
+Q24: If a CPU has 4 cores, how many processes can it execute concurrently?
+
+A CPU with 4 cores can run as many processes "Concurrently" as it has memory to store the not currently executing processes.
+
+Q25: Explain the difference between synchronous and asynchronous code. When would you prefer one technique over the other?
+
+Synchronous code has to be sequentially completed before the next set of instructions can be processed, asynchronous code allow an instruction set to be hung and another started and then returned back to the previous and so on.
+
+You would use asynchronous techniques where you can run two tasks simultaneously, usually when they do not share dependencies and are independant, and you would use synchronous techniques, when you need a task to complete before proceeding.
+
+Q26: Can a linux web-server operate acceptably with 1CPU and 512MB RAM? Can a Windows server do the same?
+
+Although Linux can run on very little RAM as little as 8MB, the act of serving as a webserver typically requires as much as 2GB RAM to functional minimally, depending of course of the load.
+The same is true of windows, except that the native cost of running windows is higher than linux and typically requires as much as 512MB of RAM to function at all.
+
+Q27: Is it bad for an operating system to have assigned all its available memory? When might this occur and why can it be beneficial to performance rather than detrimental?
+
+If an operating system assigns all of its avialable memory then there is no memory remaining for additional processes, namely memory required for memory allocation processes like dynamic memory allocation.
+Things like garbage collection of memory will also look to use memory to function and if there is no memory to assign then it would crash. 
+
+I can find no example of any instance where performance would be in any way shape or form increased by having all memory assigned. almost all memory allocation methods seem to require space to function correctly.
+
+Q28: What are the operational pros/cons of using a Linux server vs a Windows server?
+
+Linux 
+Pros:
+Greater freedom to access and configure the core components of the system, but require the knowledge to do so.
+Easier for multiple ops to work with simultaneously.
+fewer security issues
+less hardware demanding
+better remote function for remote admin
+source code available
```

Comment:
First class docker support (we haven't covered docker yet, but windows is very much a second class citizen in its ecosystem).

While it is more complicated to run, there are more resources available online to help you.